examples/multi-gpu-movielens/01-03b-MultiGPU-Download-Convert-ETL-with-NVTabular-Training-with-PyTorch.ipynb

```diff
@@ -0,0 +1,683 @@
+   "id": "22ef4df7",
+   "metadata": {},
+   "source": [
+    "#### Using Horovod with PyTorch and NVTabular\n",
+    "\n",
+    "The training script below is based on [getting-started-movielens/03a-Training-with-PyTorch.ipynb](../getting-started-movielens/03a-Training-with-PyTorch.ipynb), with a few important changes:\n",
+    "\n",
+    "- We provide several additional parameters to the `TorchAsyncItr` class, including the total number of workers `hvd.size()`, the current worker's id number `hvd.rank()`, and a function for generating random seeds `seed_fn()`. \n",
+    "\n",
+    "```python\n",
+    "    train_dataset = TorchAsyncItr(\n",
+    "        ...\n",
+    "        global_size=hvd.size(),\n",
+    "        global_rank=hvd.rank(),\n",
+    "        seed_fn=seed_fn,\n",
+    "    )\n",
+    "```\n",
+    "- The seed function uses Horovod to collectively generate a random seed that's shared by all workers so that they can each shuffle the dataset in a consistent way and select partitions to work on without overlap. The seed function is called by the dataloader during the shuffling process at the beginning of each epoch:\n",
+    "\n",
+    "```python\n",
+    "    def seed_fn():\n",
+    "        max_rand = torch.iinfo(torch.int).max // hvd.size()\n",
+    "\n",
+    "        # Generate a seed fragment\n",
+    "        seed_fragment = cupy.random.randint(0, max_rand)\n",
+    "\n",
+    "        # Aggregate seed fragments from all Horovod workers\n",
+    "        seed_tensor = torch.tensor(seed_fragment)\n",
+    "        reduced_seed = hvd.allreduce(seed_tensor, name=\"shuffle_seed\", op=hvd.mpi_ops.Sum)\n",
+    "\n",
+    "        return reduced_seed % max_rand\n",
+    "```\n",
+    "\n",
+    "- We wrap the PyTorch optimizer with Horovod's `DistributedOptimizer` class and scale the learning rate by the number of workers:\n",
+    "\n",
+    "```python\n",
+    "    optimizer = torch.optim.Adam(model.parameters(), lr=0.01 * lr_scaler)\n",
+    "    optimizer = hvd.DistributedOptimizer(optimizer, named_parameters=model.named_parameters())\n",
+    "```\n",
+    "\n",
+    "- We broadcast the model and optimizer parameters to all workers with Horovod:\n",
+    "\n",
+    "```python\n",
+    "    hvd.broadcast_parameters(model.state_dict(), root_rank=0)\n",
+    "    hvd.broadcast_optimizer_state(optimizer, root_rank=0)\n",
+    "```\n",
+    "\n",
+    "The rest of the script is the same as the MovieLens example in [getting-started-movielens/03a-Training-with-PyTorch.ipynb](../getting-started-movielens/03a-Training-with-PyTorch.ipynb). In order to run it with Horovod, we first need to write it to a file."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 10,
+   "id": "35bd70b1",
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Writing ./trainer.py\n"
+     ]
+    }
+   ],
+   "source": [
+    "%%writefile './trainer.py'\n",
```

Comment:
nice!

Reply:
This was @bschifferer's suggestion, and a really good one!